### PR TITLE
Getting rid of all the screen scraping for Xen domain status

### PR DESCRIPTION
**Pull request by rvs**
So admittedly this is a bit premature (since I wanted to get a bit more time to think how we can do an equivalent of kubernetes/docker liveness check across the board) but since we seem to be running into xl info screen scraping coredumping on weird xl info output -- here it is.

So on one hand -- this is exactly what it is -- just fixing all the brittle screen scraping Go code with more robust screen scraping shell code (although I haven't really tested it -- so if we decide to pull it in as-is -- I need a bit of time to run some basic sanity checks).

However, it is also an invitation for a discussion of how we can do the liveness check for all the different tasks that are running on EVE.

The proposed idea here is that just like Kubernetes with https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/ or Docker with https://docs.docker.com/engine/reference/builder/#healthcheck statement in a Docker file we will allow our tasks to have a declaration of "here's how I want my fine grained state to be probed".

And therein lies the rub -- we have a choice to make. On one hand we can follow Docker/k8s example and simply make it a here's the executable -- run it and get me the result OR we can actually use the filesystem approach so that whatever is running in the container will write locally to the /dev/status file and then the EVE infrastructure will pick it up from there (likely via the bind mount).

So that's choice #1. Choice #2 is standardizing on keywords (right now we return running|paused|halting|broken) but may be we need more or less of those.

@deitch @eriknordmark lets discuss where do we take this tomorrow.

### CODE REVIEW

File: pkg/pillar/hypervisor/xen.go [hunk 1]
```diff
@@ -437,46 +437,16 @@ func (ctx xenContext) Info(domainName string, domainID int) (int, types.SwState,
 	log.Infof("xen-info done. Result %s\n", stdOut)
 
 	//stdoutStderr should have 2 rows separated by '\n'. Where 1st row will be column names and 2nd row will be domain details
```
Comment (eriknordmark) [inline pkg/pillar/hypervisor/xen.go:439]: This is a leftover comment.

**Comment by deitch** [comment]
> Don't know whether we want a separate "failed" state for this and have "broken" be about crashed/device model crashed

Yes. There should be a difference between "I tried, I broke" and "I couldn't even start". Of course, if there is no "broken" other than "broke at start time", then it is the same.

> What I wonder about is catching exit codes/reasons for containers. I asumuming exit(0) vs. non-zero is something folks care about today but I could be wrong.
With that assumption, do we have a way to get that exit value.
If we do, then would it make sense to put that in a /dev/exit-reason file (similar to your /dev/status)?

Definitely do. Do we even have a similar concept for full-on VMs? Sure, their "init command" is the kernel and then the init, but do we have some concept of "and once you boot up, run this command"? Or is that opaque to us, since one might do nothing, another might use systemd, another might have a custom `init`, etc.?

I think this is a broader discussion (but we should start here narrow): is there a concept of a "core" or "primary" process and a method for it to flag desired errors/exit? For containers it is easy, as there is an explicit `ENTRYPOINT` and `CMD`. For VMs it is harder, because your init process can have many children, and which one is important, and is it opaque, and should it be? It is a much harder problem to solve, but worth spending time on separately, I think.

**Comment by deitch** [comment]
> On one hand we can follow Docker/k8s example and simply make it a here's the executable -- run it and get me the result OR we can actually use the filesystem approach so that whatever is running in the container will write locally to the /dev/status file and then the EVE infrastructure will pick it up from there (likely via the bind mount).

Two problems with the filesystem approach:

1. That requires people to change their builds, possibly their apps, does it not? Take the simple nginx example. nginx doesn't really have a way to update a file and say, "I am fine. I am fine. I am fine..." every 30 seconds or whatever. It can be built in, but that is overhead, and it is a cross-cutting concern, and then you have to pass frequency, and... I expect people will be resistant to change their apps, and many will use standard containers, and at the least expect those that work with other orchestration.
1. Do we want to swim against the tide? If people expect the common style, then unless we have something with very obvious advantages, do we want to go against it?

The two most common methods are:

* network endpoint, http/tcp/udp (supported by k8s as listed above, AWS ELB, etc.)
* command to execute

This is just what people expect.

If we _also_ want to add support for a file that gets updated, we can, but that is additional work to support and maintain it. 

Either way, if we actually were to ask people to change their apps, wouldn't we just want them to add a file to execute or a network endpoint?

**Comment by eriknordmark** [comment]
> However, it is also an invitation for a discussion of how we can do the liveness check for all the different tasks that are running on EVE.

Suppose we wanted to take the verifier and run it as a separate task, then we need to support two things to provide the same functionality as today:
1. the verifier process exits. We use the pid file for this today with watchdog wataching; as a task the task would presumably exit.
2. the verifier process runs but some event loop/goroutine is stuck. We use zero or more .touch files for this today to check liveness.

For EVE tasks we could continue to use files for #2, or switch to something else with the same flexibility of zero or more liveness indicators per task.

**Comment by deitch** [comment]
> Suppose we wanted to take the verifier and run it as a separate task

I thought this discussion was about customer workload liveness?

**Comment by deitch** [comment]
> Suppose we wanted to take the verifier and run it as a separate task

Hmm, so rather than a long-running verifier, we would just have a dispatcher? And that dispatcher would kick off verifier when it saw such a task? And the verifier process would be short-lived?

**Comment by deitch** [comment]
Sort of a FaaS/Lambda architecture?

**Comment by eriknordmark** [comment]
> I thought this discussion was about customer workload liveness?

I interpreted @rvs text as also wanting to think about EVE's microservices in this discussion.
Note that they can still be long-running tasks as now, but if we have an approach to detect exit and lack of liveness it means that we can easily run them as separate tasks with isolation between them; just need to set up the bind mounts for each agent with the correct RW/RO permissions.

**Comment by deitch** [comment]
Anything that simplifies it has my vote.

**Comment by rvs** [comment]
> But if xl create fails in xen-start we shouldn't bail "Couldn't find domain " but instead immedaitely mark it as broken. 

This is effectively what's going to happen with #1249 now @eriknordmark 

A somewhat bigger question is how do we attach a string explaining the failure to a domain that was declared broken. As we've discussed yesterday -- I plan to push in the direction of our magic /dev/status file for now (but I'm going to do it in a separate PR -- so keep an eye on it today).

> Then we can have domainmgr use state==broken to post an error in DomainStatus.
> [Don't know whether we want a separate "failed" state for this and have "broken" be about crashed/device model crashed, but that is a detail to me right now.]

Yeah, that's the other bit you and @deitch convinced me about -- let me document all the existing states as-is (I'll do it as part of the original wiki proposal) and then we can decide whether we're missing any or not.

> What I wonder about is catching exit codes/reasons for containers. I asumuming exit(0) vs. non-zero is something folks care about today but I could be wrong.

Honestly, it seems that in the container-native world everything is assumed to be asynchronous and while exit code can be a useful hint -- typically you just rely on logging and telemetry to make sense of the system. In fact, the exit code most of the time is not even visible to the end-user but gets handled by container orchestration framework as a hint of whether to restart a container, etc.

> With that assumption, do we have a way to get that exit value.
> If we do, then would it make sense to put that in a /dev/exit-reason file (similar to your /dev/status)?

Yup. That's the plan for now. Expect that other PR soon (/dev/thanks-for-all-the-fish -- but maybe that is too long ;-))

> If all of those assumptions and conditions apply for containers, then for kvm/xen we could also put a string (such as the xl create failure output) into that /dev/exit-reason. Plus if kvm/xen can tell the difference between a shutdown/poweroff vs a kernel panic it could also go to the same place.

Agreed. But that'd be a subject of much further set of improvements (once we agree on the basic infrastructure for how to capture these things).

> And if the cgroup doesn't have sufficient resources to start a container do we get some distinct exit code or error from containerd? If so we could provide a more specific error than "exit code 1" in this /dev/exit-reason

I don't think containerd is of much help here. In fact, IIRC most of cgroups mechanics is expecting the caller to make sure that resources add up. In that sense it is unlike xl -- if you specified more memory than Xen has available xl will fail -- cgroups on the other hand would happily allow you to put a caps in that add up to more than physical memory on the box.

There may be one exception to that rule -- which is cgroup hierarchy -- so it would be something to consider.

Long story short, I think we need to rely on task introspection to make sure we get the correct reason of failure.

>But as you say, first let's figure out what makes sense for containers, and then decide how much work do to for xen/kvm. 
> But getting the resource errors out for xen and kvm would be useful since we don't do any resource checks before spawning the task.

+1

File: pkg/pillar/hypervisor/xen.go (hunk 1 above)
Comment (rvs) [reply pkg/pillar/hypervisor/xen.go]: Fixed.

**Comment by rvs** [comment]
Hey @deitch wanted to follow up on a few things here:

> I think this is a broader discussion (but we should start here narrow): is there a concept of a "core" or "primary" process and a method for it to flag desired errors/exit? For containers it is easy, as there is an explicit ENTRYPOINT and CMD. For VMs it is harder, because your init process can have many children, and which one is important, and is it opaque, and should it be? It is a much harder problem to solve, but worth spending time on separately, I think.

I think we can unify the backchannel between containers and VMs quite easily -- but both of these require active participation from whatever is running inside of the VM (you can lead the horse to water, but you can't make it drink so to speak ;-)). I feel like a good mid point here (which I'm going to bring onto eve-tsc@ shortly) is to simply agree on what kind of a channel will that be. A magic /dev/status /dev/exit-note seems to be a good starting point.

> Two problems with the filesystem approach:
> That requires people to change their builds, possibly their apps, does it not? Take the simple nginx example. nginx doesn't really have a way to update a file and say, "I am fine. I am fine. I am fine..." every 30 seconds or whatever. It can be built in, but that is overhead, and it is a cross-cutting concern, and then you have to pass frequency, and... I expect people will be resistant to change their apps, and many will use standard containers, and at the least expect those that work with other orchestration.

While that is true, it is certainly WAY easier in my book than running a process inside of a VM (in fact, unlike containers you simply CAN'T do that with VMs -- so that leaves only network endpoint -- see below).

> Do we want to swim against the tide? If people expect the common style, then unless we have something with very obvious advantages, do we want to go against it?

I was thinking exactly along the same line until the lightbulb went on during our chat -- since we need to collect postmortem information from tasks that have exited -- we are really stuck with having some kind of a communication channel that is effectively push (from the task point of view) AND can survive domain dying. Neither execing the process nor hitting an HTTP endpoint (and such) inside the task satisfy that criteria. Hence we need to do magic /dev/X anyway. And once we do -- we may as well use it for other purposes.

> Either way, if we actually were to ask people to change their apps, wouldn't we just want them to add a file to execute or a network endpoint?

Lets start small -- lets solve the issue in front of us (that seems to require /dev/X anyway) and then we can open it up for more ways of communicating extra status.

Makes sense @deitch ?
